### PR TITLE
`Development`: Enable parallel server test execution

### PR DESCRIPTION
**Pull request by DominikRemo**
#### General

- [x] The changed server tests succeed **locally** and on **Bamboo** and on **GitHub Actions**.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

### Motivation and Context
Fast tests 🏎️
Improves server-test execution time from up to 30 min 🚗 => down to 12 min 🏎️

### Description
Enables parallel testing with JUnit 5 [JUnit Docs: 2.20. Parallel Execution](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parallel-execution).

Implements a new [JUnit 5 extension](https://junit.org/junit5/docs/current/user-guide/#extensions) and custom [Logback appender](https://logback.qos.ch/manual/appenders.html) for logging in parallel tests.

Fixes several flaky tests (e.g.: in classes LongFeedbackResourceIntegrationTest, NotificationScheduleServiceTest and ProgrammingExerciseScheduleServiceTest)

Re-organized tests into different groups for better balancing.

- [x] We confirm that the tests on this branch are at least as reliable as on the develop branch. We tested the reliability of the tests with a custom build plan (_development-enable-parallel-unit-testing_ on https://bamboo.ase.in.tum.de/browse/ARTEMIS-STO0).
We used a custom plan that only executes the server tests. Besides only running the server tests, the _Server Tests Only_ plan is similar to the _Artemis Test_ plan

### Steps for Testing
**Test changes only**

#### Code Review
- [x] Code Review 1
- [x] Code Review 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/scheduled/NotificationScheduleService.java [hunk 1]
```diff
@@ -164,8 +164,8 @@ private void scheduleNotificationForAssessedExercisesSubmissions(Exercise exerci
      * @return true if the time is valid else false
      */
     private boolean checkIfTimeIsCorrectForScheduledTask(ZonedDateTime relevantTime) {
-        // only send a notification if relevantTime is defined and not in the future (i.e. in the range [now-2 minutes, now]) (due to possible delays in scheduling)
-        return relevantTime != null && !relevantTime.isBefore(ZonedDateTime.now().minusMinutes(2)) && !relevantTime.isAfter(ZonedDateTime.now());
+        // only send a notification if relevantTime is defined and not in the future (i.e. in the range [now-2 minutes, now+2 minutes]) (due to possible delays in scheduling)
+        return relevantTime != null && !relevantTime.isBefore(ZonedDateTime.now().minusMinutes(2)) && !relevantTime.isAfter(ZonedDateTime.now().plusMinutes(2));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/scheduled/NotificationScheduleService.java:168]: Why is this change necessary? Also the comment mentioning "not in the future" and "now + 2 minutes" does not make sense to me.

Comment (DominikRemo) [reply src/main/java/de/tum/in/www1/artemis/service/scheduled/NotificationScheduleService.java:168]: Maybe we should change the comment from "not in the future" to "not in the far future"?

**Review by Strohgelaender — DISMISSED** [review]
code lgtm, tests are much faster now 🚀 thanks for implemeting!

**Review by terlan98 — COMMENTED** [review]
Well done! The tests are indeed a lot faster now 🚀 
The code looks good, but I have one question:

**Review by terlan98 — DISMISSED** [review]

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java [hunk 2]
```diff
@@ -35,6 +38,8 @@ class DatabaseQueryCountTest extends AbstractSpringIntegrationBambooBitbucketJir
     void setup() {
         participantScoreScheduleService.shutdown();
         userUtilService.addUsers(TEST_PREFIX, 1, NUMBER_OF_TUTORS, 0, 0);
+        User student = userUtilService.getUserByLogin(TEST_PREFIX + "student1");
+        student.setGroups(Set.of(TEST_PREFIX + "tumuser"));
```
[the remark this answers is not in the document]
Comment (laadvo) [reply src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java:42]: The test was mostly flaky locally. In cases where it was run towards the end, there were some other courses the student was a part of due to the default groups. This caused the test to fail due to too many database calls. The overhead of changing the groups after user creation is minimal and the test did not seem to be flaky afterwards :)
I hope this answers your question.

[the remark this answers is not in the document]
Comment (terlan98) [reply src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java:42]: Thanks for the detailed answer 👍

File: src/test/java/de/tum/in/www1/artemis/notification/NotificationResourceIntegrationTest.java [hunk 3]
```diff
@@ -61,6 +61,7 @@ void initTestCase() {
         course1 = textExerciseUtilService.addCourseWithOneReleasedTextExercise();
         course2 = textExerciseUtilService.addCourseWithOneReleasedTextExercise();
         systemNotificationRepository.deleteAll();
+        notificationRepository.deleteAll();
```
[the remark this answers is not in the document]
Comment (DominikRemo) [reply src/test/java/de/tum/in/www1/artemis/notification/NotificationResourceIntegrationTest.java:60]: Yes it does. Thank you for noticing :)
It's fixed by removing the _systemNotificationRepository.deleteAll()_ call in two places

**Review by MaximilianAnzinger — DISMISSED** [review]
Thank you for considering my comment. Hopefully, this gets merged asap 👍

**Review by terlan98 — DISMISSED** [review]

**Review by Strohgelaender — DISMISSED** [review]

**Pull request merged**